README.md
![Nodezoo][Banner]

# nodezoo-travis
- __Lead Maintainer:__ [Matthew O'Connor][Lead]
- __Sponsor:__ [nearForm][Sponsor]

A micro-service that provides Travis data for [NodeZoo][]. This micro-service depends
on the Travis and NPM registries but also caches retrieved data to reduce load on both
public registries.

If you're using this microservice, and need help, you can:

- Post a [github issue][],
- Tweet to [@nodezoo][],
- Ask on the [Gitter][gitter-url].

## Running
This micro-service can be ran as part of a complete system or as a single isolated
unit.

### As a complete system
A special system repository is available that runs the complete system using Docker
and Fuge.

- [Nodezoo: The complete system][System]

### Isolated mode
To make testing easier this micro-service can be ran in 'isolated' mode. This mode
allows testing over http using a well defined port. Please note isolated mode means
patterns are not exposed via mesh.

To run in isolated mode,

 - Clone this repository locally,
 - Run `npm install`,
 - Run `npm start isolated`,

__Note:__ In memory storage is used over redis in isolated mode.

A simple http service is supported and can be called using Curl or other Rest client.
The default port is `8053`. It can be changed using the `TRAVIS_PORT` environment
variable.

```
curl -d '{"role":"travis","cmd":"get","name":"hapi"}' http://localhost:8052/act
```

## Configuration

### Environment Variables
Various settings can be changed using environment variables, see the list below for
all available variable names.

#### TRAVIS_HOST
  - The host to listen on in isolated mode.
  - Defaults to `localhost`

#### TRAVIS_PORT
  - The port to listen on in isolated mode.
  - Defaults to `8051` .

#### TRAVIS_REDIS_HOST
  - The host redis will listen on.
  - Defaults to `localhost`

#### TRAVIS_REDIS_PORT
  - The port redis listen on.
  - Defaults to `6379` .

#### TRAVIS_ISOLATED
  - Starts isolated mode.
  - Defaults to `false`.

#### TRAVIS_REGISTRY
  - Change the npm registry used to validate the module name.
  - Defaults to `http://registry.npmjs.org/`.

## Sample Data
```json
{
  "entity$": "-/-/travis_cache",
  "name": "seneca-web",
  "url": "https://travis-ci.org/senecajs/seneca-web",
  "buildId": "1398674",
  "active": "true",
  "buildState": "passed",
  "lastBuilt": "2016-02-29T17:05:48Z",
  "cached": "1460328882872",
  "id": "seneca-web"
}
```

## Messages Handled

### `role:travis,cmd:get`
Returns Travis specific data for the module name provided.

```js
seneca.act(`role:travis,cmd:get`, {name:'seneca'}, (err, data) => {})
```

### `role:info,req:part`
An alias for `role:travis,cmd:get`, allows integration into the wider nodezoo-system.

```js
seneca.act(`role:info,req:part`, {name:'seneca'}, (err, reply) => {})
```

## Messages Emitted

### `role:info,res:part`

Called in response to a call to `role:info,req:part`.

```js
seneca.add(`role:info,res:part`, (msg, done) => {})
```

## Data Emitted
- name - Name of the module
- url - URL to the module on the Travis-CI website
- buildId - ID
- active - Whether the repository is active or not
- buildState - It's most recent build state
- lastBuilt - The time and date the module was last built at
- cached - The time the data was last cached at

## Contributing
The [Nodezoo][Org] encourages __open__ and __safe__ participation.

- [Code of Conduct][CoC]

If you feel you can help in any way, be it with documentation, examples, extra testing, or new
features please get in touch.

## License
Copyright (c) 2016, Matthew O'Connor and other contributors.
Licensed under [MIT][].

[Banner]: https://raw.githubusercontent.com/nodezoo/nodezoo-org/master/assets/logo-nodezoo.png
[Lead]: https://github.com/mcdonnelldean
[Sponsor]: http://www.nearform.com/
[Org]: https://github.com/nodezoo
[CoC]: https://github.com/nodezoo/nodezoo-org/blob/master/CoC.md
[MIT]: ./LICENSE
[github issue]: https://github.com/nodezoo/nodezoo-npm/issues
[@nodezoo]: http://twitter.com/nodezoo
[gitter-url]: https://gitter.im/nodezoo/nodezoo-org
[System]: https://github.com/nodezoo/nodezoo-system
[NodeZoo]: https://github.com/rjrodger/nodezoo


CHANGELOG.md
## 10 April 2016 - 7.0.0

* **BREAKING** Data structure changed.
* **BREAKING** group state deleted.
* Added CHANGELOG.md.
* Results set refactored.
* tavis.js and start.js refactored.
* Added isolated mode.
* Updated README to reflect changes.
* Added sample data output to README.md.
* Removed unneeded call to travis.
* Removed testing logs
* Fixed crash on build array being empty
* Removed toolbag

## 6 April 2016 - 6.0.0

* First release of the 6th major revision of nodezoo.
* Switched to using npm released packages instead of git checkouts.


test.js
var t = require('./travis.js')

 t.cmd_get({input:10380000})

test/index.test.js
'use strict'

var Code = require('code')
var Lab = require('lab')
var Seneca = require('seneca')
var _ = require('lodash')

var lab = exports.lab = Lab.script()
var describe = lab.describe
var suite = lab.suite
var before = lab.before
var it = lab.it
var expect = Code.expect

function createInstance () {
  return Seneca({log: 'silent'})
    .use('entity')
    .use('../lib/travis')
}
process.setMaxListeners(12)

var si = createInstance()

suite('nodezoo-travis suite tests ', function () {
  before({}, function (done) {
    si.ready(function (err) {
      if (err) {
        return process.exit(!console.error(err))
      }
      done()
    })
  })
})

describe('nodezoo-travis tests', () => {
  it('test Valid Response', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply).to.not.be.empty()
      done(err)
    })
  })
  it('test Everything', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply).to.not.be.empty()
      done(err)
    })
  })
  it('test No Repo', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': '' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply).to.not.be.empty()
      done(err)
    })
  })
  it('test no User', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': '', 'repo': '' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply).to.not.be.empty()
      done(err)
    })
  })
  it('test Invalid Response', function (done) {
    var si = createInstance()
    var payload = { 'name': 'shoobydoobydoobop' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.exist()
      done()
    })
  })
  it('test ID exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.id).to.be.a.string()
      done(err)
    })
  })
  it('test URL exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.url).to.be.a.string()
      done(err)
    })
  })
  it('test BuildID exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.buildId).to.be.a.number()
      done(err)
    })
  })
  it('test Active exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.active).to.be.a.boolean()
      done(err)
    })
  })
  it('test buildState exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.buildState).to.be.a.string()
      done(err)
    })
  })
  it('test lastBuilt exists', function (done) {
    var si = createInstance()
    var payload = { 'name': 'seneca', 'user': 'powerbrian', 'repo': 'https://github.com/senecajs/seneca' }
    si.act(_.extend({ role: 'travis', cmd: 'get' }, payload), function (err, reply) {
      expect(err).to.not.exist()
      expect(reply.lastBuilt).to.be.a.string()
      done(err)
    })
  })
})


.toolbagrc.js
'use strict'

const StatsCollector = require('toolbag/lib/plugins/toolbag_plugin_stats_collector')
const UdpReporter = require('toolbag/lib/plugins/toolbag_plugin_udp_reporter')

module.exports = function (defaults, callback) {
  callback(null, {
    plugins: [
      {
        plugin: UdpReporter,
        options: {
          id: 'udp reporter',
          socketType: 'udp4',
          port: 5001,
          host: 'localhost'
        }
      },
      {
        plugin: StatsCollector,
        options: {
          enabled: true,
          period: 1000,
          eventLoopLimit: 30,
          features: {
            process: true,
            system: true,
            cpu: true,
            memory: true,
            gc: false,
            handles: false,
            requests: false,
            eventLoop: true,
            meta: {
              tags: [
                'nodezoo-travis'
              ]
            }
          }
        }
      }
    ]
  })
}


lib/travis.js
'use strict'

var Travis = require('travis-ci')
var Request = require('request')

var opts = {
  registry: 'http://registry.npmjs.org/'
}

module.exports = function () {
  var seneca = this

  opts = seneca.util.deepextend(opts)
  opts.client = new Travis({version: '2.0.0'})
  opts.cache = seneca.make$('travis_cache')
  opts.cache.load$()

  seneca.add('role:travis,cmd:get', cmdGet)
  seneca.add('role:info,req:part', aliasGet)

  return {
    name: 'nodezoo-travis'
  }
}

function cmdGet (msg, done) {
  var cache = opts.cache
  var registry = opts.registry + msg.name

  cache.load$(msg.name, (err, cached) => {
    if (err) return done(err)

    if (cached && !msg.update) {
      return done(null, cached.data$(cached))
    }

    Request.get({url: registry, gzip: true}, (err, res, body) => {
      if (err) return done(err)

      var data = JSON.parse(body)
      var distTags = data['dist-tags'] || {}
      var latest = ((data.versions || {})[distTags.latest]) || {}
      var repository = latest.repository || {}
      var url = repository.url || ''

      var matches = /[\/:]([^\/:]+?)[\/:]([^\/]+?)(\.git)*$/.exec(url) || []
      var params = {
        name: msg.name,
        user: matches[1] || '',
        repo: matches[2] || '',
        cached: cached
      }

      queryTravis(params, done)
    })
  })
}

function queryTravis (msg, done) {
  var cache = opts.cache
  var client = opts.client

  var user = msg.user
  var repo = msg.repo

  client.repos(user, repo).get((err, res) => {
    if (err) return done(err)

    var build = res.repo || {}
    var data = {
      name: msg.name || '',
      url: 'https://travis-ci.org/' + user + '/' + repo,
      buildId: build.id || '',
      active: build.active || '',
      buildState: build.last_build_state || '',
      lastBuilt: build.last_build_started_at || '',
      cached: Date.now()
    }

    function complete (err, entity) {
      if (err) return done(err)
      else done(null, entity.data$(entity))
    }

    if (msg.cached) {
      msg.cached.data$(data).save$(complete)
    }
    else {
      data.id$ = msg.name
      cache.make$(data).save$(complete)
    }
  })
}

function aliasGet (msg, done) {
  var seneca = this
  var payload = {name: msg.name}

  seneca.act('role:travis,cmd:get', payload, (err, data) => {
    if (err) return done(err)

    payload.data = data
    seneca.act('role:info,res:part,part:travis', payload)
    done()
  })
}


srv/start.js
'use strict'

var Seneca = require('seneca')
var Entities = require('seneca-entity')
var Mesh = require('seneca-mesh')
var Travis = require('../lib/travis')
var RedisStore = require('seneca-redis-store')

var envs = process.env
var opts = {
  seneca: {
    tag: envs.TRAVIS_TAG || 'nodezoo-travis'
  },
  travis: {
    token: envs.TRAVIS_TOKEN || 'NO_TOKEN',
    registry: envs.TRAVIS_REGISTRY || 'http://registry.npmjs.org/'
  },
  mesh: {
    auto: true,
    listen: [
      {pin: 'role:travis,cmd:get', model: 'consume'},
      {pin: 'role:info,req:part', model: 'observe'}
    ]
  },
  isolated: {
    host: envs.TRAVIS_HOST || 'localhost',
    port: envs.TRAVIS_PORT || '8053'
  },
  redis: {
    host: envs.TRAVIS_REDIS_HOST || 'localhost',
    port: envs.TRAVIS_REDIS_PORT || '6379'
  }
}

var Service = Seneca(opts.seneca)

Service.use(Entities)

if (envs.TRAVIS_ISOLATED) {
  Service.listen(opts.isolated)
}
else {
  Service.use(Mesh, opts.mesh)
  Service.use(RedisStore, opts.redis)
}

Service.use(Travis, opts.travis)
